### PR TITLE
net/ieee802154/radio: use bitflags for capabilities

### PR DESCRIPTION
**Pull request by jeandudey**
### Contribution description

Modifies the `ieee802154_radio_ops_t` structure to use an integer to store capabilities instead of a function call. Uses an `uint16_t` for this purpose, in the future, if needed could be extended to an `uint32_t` if more capabilities are needed.

The `caps` field was put on the top of the structure just for aesthetics reasons (to not mix actual operations with the capabilities).

The motivation is to (on another PR) add the PHY capabilities, see https://github.com/RIOT-OS/RIOT/pull/15932#discussion_r570904375 and previous comment on the HAL RDM (https://github.com/RIOT-OS/RIOT/pull/13943#discussion_r414769629).

<details>
<summary>Binary size changes</summary>

Tested with `make -C examples/gnrc_networking BOARD=<....>`

#### PR

**cc2538dk**
```
   text	   data	    bss	    dec	    hex	filename
  91512	    208	  19092	 110812	  1b0dc	/home/jeandudey/Dev/RIOT/examples/gnrc_networking/bin/cc2538dk/gnrc_networking.elf
```

**nrf52840dk**

```
   text	   data	    bss	    dec	    hex	filename
  88796	    200	  19388	 108384	  1a760	/home/jeandudey/Dev/RIOT/examples/gnrc_networking/bin/nrf52840dk/gnrc_networking.elf
```

#### master

**cc2538dk**
```
   text	   data	    bss	    dec	    hex	filename
  91556	    208	  19092	 110856	  1b108	/home/jeandudey/Dev/RIOT/examples/gnrc_networking/bin/cc2538dk/gnrc_networking.elf
```

**nrf52840dk**

```
   text	   data	    bss	    dec	    hex	filename
  88836	    200	  19388	 108424	  1a788	/home/jeandudey/Dev/RIOT/examples/gnrc_networking/bin/nrf52840dk/gnrc_networking.elf
```

The `text`  section was reduced by 44 bytes on `cc2538dk` and 40 bytes on `nrf52840dk`. `data` and `bss` stayed the same.

</details>

### Testing procedure

- `make -C examples/gnrc_networking BOARD=cc2538dk`
- `make -C examples/gnrc_networking BOARD=nrf52840dk`

*Note:* i don't own any of the boards for actual testing, so if something is wrong or doesn't work as expected leave a comment 😉 .

### Issues/PRs references

See #15932

### CODE REVIEW

**Comment by jia200x** [comment]
PR 16000 :tada:

File: sys/include/net/ieee802154/radio.h [hunk 1]
```diff
@@ -1056,14 +1054,14 @@ static inline bool ieee802154_radio_has_auto_csma(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_sub_ghz(ieee802154_dev_t *dev)
 {
-    return dev->driver->get_cap(dev, IEEE802154_CAP_SUB_GHZ);
+    return (dev->driver->caps & IEEE802154_CAP_SUB_GHZ) != 0;
```
Comment (jia200x) [inline sys/include/net/ieee802154/radio.h:1057]: for all these functions, it should be enough to do:

```suggestion
    return (dev->driver->caps & IEEE802154_CAP_SUB_GHZ);
```

File: sys/include/net/ieee802154/radio.h [hunk 2]
```diff
@@ -382,6 +383,14 @@ typedef struct {
  * @brief Radio ops struct declaration
  */
 struct ieee802154_radio_ops {
+    /**
+     * @brief Radio device capabilities
+     *
+     * This field contains bitflags of supported capabilities
+     * (@ref ieee802154_rf_caps_t) by the device.
+     */
+    uint16_t caps;
```
Comment (jia200x) [inline sys/include/net/ieee802154/radio.h:392]: since this won't change, it can be declared as `const`
```suggestion
    const uint16_t caps;
```

**Comment by jia200x** [comment]
Some small comments. Otherwise I'm fine with this PR!

**Comment by jia200x** [comment]
For the `nrf802154`:
```
2021-02-15 11:07:39,411 #                CAPS              
2021-02-15 11:07:39,414 # =================================
2021-02-15 11:07:39,417 # - Frame Retransmissions:        n
2021-02-15 11:07:39,420 #     * Info retries:             n
2021-02-15 11:07:39,423 # - Auto CSMA-CA:                 n
2021-02-15 11:07:39,426 # - 2.4 GHz band:                 y
2021-02-15 11:07:39,429 # - SubGHz band:                  n
2021-02-15 11:07:39,432 # - TX DONE indication:           y
2021-02-15 11:07:39,435 #     * ACK Timeout indication:   n
2021-02-15 11:07:39,438 # - RX_START indication:          y
2021-02-15 11:07:39,441 # - CCA Done indication:          y
```

For the cc238-rf:
```
2021-02-15 11:08:58,206 #                CAPS              
2021-02-15 11:08:58,207 # =================================
2021-02-15 11:08:58,209 # - Frame Retransmissions:        n
2021-02-15 11:08:58,210 #     * Info retries:             n
2021-02-15 11:08:58,211 # - Auto CSMA-CA:                 y
2021-02-15 11:08:58,213 # - 2.4 GHz band:                 y
2021-02-15 11:08:58,214 # - SubGHz band:                  n
2021-02-15 11:08:58,216 # - TX DONE indication:           y
2021-02-15 11:08:58,217 #     * ACK Timeout indication:   n
2021-02-15 11:08:58,218 # - RX_START indication:          y
2021-02-15 11:08:58,220 # - CCA Done indication:          y
```

It works!

**Comment by jia200x** [comment]
(I tried it with https://github.com/RIOT-OS/RIOT/pull/16000#discussion_r576052391. Feel free to use the patch):

```
diff --git a/sys/include/net/ieee802154/radio.h b/sys/include/net/ieee802154/radio.h
index 939ac4f906..65861fbf46 100644
--- a/sys/include/net/ieee802154/radio.h
+++ b/sys/include/net/ieee802154/radio.h
@@ -1006,7 +1006,7 @@ static inline int ieee802154_radio_confirm_cca(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_irq_ack_timeout(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_IRQ_ACK_TIMEOUT) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_IRQ_ACK_TIMEOUT);
 }
 
 /**
@@ -1022,7 +1022,7 @@ static inline bool ieee802154_radio_has_irq_ack_timeout(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_frame_retrans(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_FRAME_RETRANS) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_FRAME_RETRANS);
 }
 
 /**
@@ -1038,7 +1038,7 @@ static inline bool ieee802154_radio_has_frame_retrans(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_auto_csma(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_AUTO_CSMA) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_AUTO_CSMA);
 }
 
 /**
@@ -1054,7 +1054,7 @@ static inline bool ieee802154_radio_has_auto_csma(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_sub_ghz(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_SUB_GHZ) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_SUB_GHZ);
 }
 
 /**
@@ -1070,7 +1070,7 @@ static inline bool ieee802154_radio_has_sub_ghz(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_24_ghz(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_24_GHZ) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_24_GHZ);
 }
 
 /**
@@ -1086,7 +1086,7 @@ static inline bool ieee802154_radio_has_24_ghz(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_irq_tx_done(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_IRQ_TX_DONE) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_IRQ_TX_DONE);
 }
 
 /**
@@ -1102,7 +1102,7 @@ static inline bool ieee802154_radio_has_irq_tx_done(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_irq_rx_start(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_IRQ_RX_START) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_IRQ_RX_START);
 }
 
 /**
@@ -1118,7 +1118,7 @@ static inline bool ieee802154_radio_has_irq_rx_start(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_irq_tx_start(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_IRQ_TX_START) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_IRQ_TX_START);
 }
 
 /**
@@ -1134,7 +1134,7 @@ static inline bool ieee802154_radio_has_irq_tx_start(ieee802154_dev_t *dev)
  */
 static inline bool ieee802154_radio_has_irq_cca_done(ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_IRQ_CCA_DONE) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_IRQ_CCA_DONE);
 }
 
 /**
@@ -1152,7 +1152,7 @@ static inline bool ieee802154_radio_has_irq_cca_done(ieee802154_dev_t *dev)
 static inline bool ieee802154_radio_has_frame_retrans_info(
     ieee802154_dev_t *dev)
 {
-    return (dev->driver->caps & IEEE802154_CAP_FRAME_RETRANS_INFO) != 0;
+    return (dev->driver->caps & IEEE802154_CAP_FRAME_RETRANS_INFO);
 }
 
 /**
```

**Comment by jeandudey** [comment]
Thanks for the review @jia200x ! Fixed the two suggestions 😉

**Comment by jia200x** [comment]
please squash :)

**Comment by jeandudey** [comment]
Done! 👍🏻

**Review by jia200x — APPROVED** [review]
ACK

**Comment by jia200x** [comment]
thanks for the contribution!

**Pull request merged**